libs/langchain/langchain/llms/sagemaker_endpoint.py

```diff
@@ -338,14 +338,16 @@ def _call(
                 )
                 iterator = LineIterator(resp["Body"])
                 current_completion: str = ""
+                start_json = b"{"
                 for line in iterator:
-                    resp = json.loads(line)
-                    resp_output = resp.get("outputs")[0]
-                    if stop is not None:
-                        # Uses same approach as below
-                        resp_output = enforce_stop_tokens(resp_output, stop)
-                    current_completion += resp_output
-                    run_manager.on_llm_new_token(resp_output)
+                    if line != b"" and start_json in line:
+                        token = json.loads(
+                            line[line.find(start_json) :].decode("utf-8")
+                        )["token"]
+                        resp_output = token["text"]
+                        if not token["special"]:
+                            current_completion += resp_output
+                            run_manager.on_llm_new_token(resp_output)
```

Comment:
@lyoh001 
Thanks for working on this. I don't have the setup to verify this, but does this apply universally to all streaming model outputs that might be hosted on SM endpoints? Which specific model did you test this with?

Reply:
@3coins 
Hi Piyush.
I appreciate your feedback on my code.
I have verified the code with Llama2, Mistral, Zephyr, Falcon etc.
I have also added _acall function.
Here is a video of running against 3 different models (ie endpoints) asynchronously with fastapi.

https://www.youtube.com/watch?v=KsDXBPjDy_o

I would be glad to have a quick video call if you want :)
Please let me know if I should include _acall function in this pull request or create a new one?

Thank you.

Reply:
@lyoh001 
Thanks for sharing that info; code looks good. For the `_acall` function, I would suggest opening a separate PR.

Reply:
@lyoh001 
Thanks for testing with all those models. My concern is that the change here is different from the earlier code which was extracting from `outputs[0]` vs `token[text]` here. Do we need an output parser `streaming_transform_output` that users can define, similar to the the `transform_output` handler that can handle different model outputs?
Also, this change seemed to have omitted the `enforce_stop_tokens(resp_output, stop)` statement.

Reply:
@lyoh001 
I looked at the linked SageMaker post and it seems like there are 2 different container types that support streaming response, and output in these differ. Here is the excerpt from the blog you referred to.

> In this post, we showcase two container options to create a SageMaker endpoint with response streaming: using an AWS [Large Model Inference](https://docs.aws.amazon.com/sagemaker/latest/dg/large-model-inference-dlc.html) (LMI) and Hugging Face [Text Generation Inference](https://huggingface.co/text-generation-inference) (TGI) container

LMI seems to use the existing `outputs[0]` and TGI uses `token[text]`. I believe we need both including the new implementation that you have included here to support both of these containers. Let me know if I am missing anything.

Reply:
@3coins 

Hi Piyush,

Thank you for your feedback. I agree that the different output formats from LMI and TGI containers need to be handled properly. I will update the _call and _acall functions to include a parameter for the output parser, as you suggested. I will run some tests to validate the changes and update this pull request soon. I appreciate your review and suggestions.

Reply:
@lyoh001 
Can the two outputs be auto-detected in any way? For example:
```python
try:
    resp = json.loads(line)
    resp_output = resp.get("outputs")[0]
except JSONDecodeError:
    // TGI code
except KeyError:
   // TGI code
```

Reply:
@3coins 

Hi Piyush,

Thank you for your feedback. I have found a way to detect the output format from LMI or TGI containers based on the HTTP header “x-amzn-sagemaker-content-type”. According to the blog (and I confirmed), TGI returns “text/event-stream” while LMI returns “application/jsonlines”. Therefore, we can use this header value to determine which output parser to use.
Are you ok with this approach?

![http_header](https://github.com/langchain-ai/langchain/assets/45280343/5d46e786-88e7-4a74-9467-486664443d3f)

Reply:
@lyoh001 
Excellent find 👍. Please go ahead with that approach.

Reply:
@3coins Thanks Piyush

I modified the code and also added _acall function too.
I opened up a new pull request here #12122 for that